prometheus/prometheus.go

```diff
@@ -126,11 +134,41 @@ func NewPrometheusSinkFrom(opts PrometheusOpts) (*PrometheusSink, error) {
 	return sink, reg.Register(sink)
 }
 
-// Describe is needed to meet the Collector interface.
+// Describe sends a Collector.Describe value for all gauges, summaries and counters defined on the PrometheusSink
 func (p *PrometheusSink) Describe(c chan<- *prometheus.Desc) {
-	// We must emit some description otherwise an error is returned. This
-	// description isn't shown to the user!
-	prometheus.NewGauge(prometheus.GaugeOpts{Name: "Dummy", Help: "Dummy"}).Describe(c)
+	p.gauges.Range(func(_, value interface{}) bool{
+		if value == nil {
+			return true
+		}
+		g := value.(*gauge)
+		g.Describe(c)
+
+		return true
+	})
+
+	p.summaries.Range(func(_, value interface{}) bool{
+		if value == nil {
+			return true
+		}
+		s := value.(*summary)
+		s.Describe(c)
+		return true
+	})
+
+	p.counters.Range(func(_, value interface{}) bool{
+		if value == nil {
+			return true
+		}
+		co := value.(*counter)
+		co.Describe(c)
+		return true
+	})
+
+	// dummy value to be able to register and unregister "empty" sinks
+	// Note this is not actually retained in the PrometheusSink so this has no side effects
+	// on the caller's sink. So it shouldn't show up to any of its consumers.
+	prometheus.NewGauge(prometheus.GaugeOpts{Name: p.name, Help: p.name}).Describe(c)
```

Comment:
I know this already exists, but I am surprised it is necessary because those same docs say this:

	// Sending no descriptor at all marks the Collector as “unchecked”,
	// i.e. no checks will be performed at registration time, and the
	// Collector may yield any Metric it sees fit in its Collect method.
	
That to me suggests that it would be safe to send nothing on the channel, but maybe I am interpreting it wrong.

Reply:
Thanks for the pointers -- I'll update this PR or open a new one that has just the `.Name` change.

> any ephemeral metric will change the list of descriptors, which will make it impossible to deregister

hm, just to be explicit -- I think there can be a race between deregistering a `Collector` type and modifying its underlying metrics that can make the `Unregister` fail. It wouldn't be a deterministic problem, but a racy one.

---

I took `superset` to mean the set of all possible metrics that a `Collector` collects and more. Where "more" can be the addition of the `p.Name` in the current form of the code or any other descriptor.